src/main.rs

```diff
@@ -157,3 +226,53 @@ fn parse_line(line: &str) -> anyhow::Result<Option<DVec3>> {
         _ => Ok(None),
     }
 }
+
+fn histogram(positions: &Vec<DVec3>, r_variable: fn(&DVec3) -> f64) -> (Vec<Vec<DVec3>>, Vec<f64>) {
+    let mut position_set: Vec<DVec3> = positions.clone();
+
+    let n = position_set.len();
+    let avg = position_set.iter().copied().sum::<DVec3>() / n as f64;
+    let std_dev = (position_set
+        .iter()
+        .copied()
+        .map(|r| (r - avg).powf(2.))
+        .sum::<DVec3>()
+        / (n - 1) as f64)
+        .powf(0.5);
+
+    let cutoff: i32 = 3; // measured in standard deviations
+    let div: i32 = 6;
+    let mut range = (-(cutoff * div)..(cutoff * div))
+        .into_iter()
+        .map(|i| (i as f64) / (div as f64) * r_variable(&std_dev) + r_variable(&avg))
+        .peekable();
+    position_set.sort_by(|a, b| {
+        r_variable(a)
+            .partial_cmp(&r_variable(b))
+            .expect("Histogram: Incomparable values")
+    });
+
+    let division_values = range.clone().collect::<Vec<f64>>();
+    let mut histogram: Vec<Vec<DVec3>> = Vec::new();
+    let mut division: Vec<DVec3> = Vec::new();
+
+    for pos in position_set {
+        while range.peek().is_some() {
+            if r_variable(&pos) < *range.peek().unwrap() {
+                division.push(pos);
+                break;
+            } else {
+                histogram.push(division.clone());
+                division.clear();
+                range.next();
+            }
+        }
+    }
+
+    while range.next().is_some() {
+        histogram.push(division.clone());
+        division.clear();
+    }
+    histogram.remove(0); //  Removes lower bound of data (atypical data)
```

Comment:
...And then you're removing the only element that has non-zero data, so the histogram ends up filled with zeros.

Reply:
The first element of the histogram contains every element below the first element of range, which is the lower bound of valid data (any data below this threshold is atypical and should be removed).